### DISCUSSION TITLE
An EntityAdapter implementation proposal

### DISCUSSION THREAD

**Discussion by michaeljota** [pmndrs/zustand#2402] [Ideas]
Update: You can find an implementation of this proposal here: https://www.npmjs.com/package/zustand-entity-adapter. Thanks!

---

I want to propose an EntityAdapter implementation. An EntityAdapter is described by the ngrx and the Redux Toolkit teams as a set of prebuilt methods and selectors for efficiently performing CRUD operations on a normalized state structure for a single entity state collection.

On a high level, both implementations can be divided into three parts, the state, the actions, and the selectors, and both implementations merge those parts into one adapter. Neither of the current implementations can be ported directly to Zustand because both are heavily based on their respective libraries.

Considering all this, I propose implementing a similar behavior, with a similar interface, for Zustand.

## The state

Zustand allows us to create a minimal state factory, and delegate an extra state addition to the implementer.

The initial state should have this shape:

```ts
export type EntityId = string | number;

type Dictionary<Entity, Id extends EntityId> = Partial<Record<Id, Entity>>;

export interface EntityState<Entity, Id extends EntityId> {
  ids: Id[];
  entities: Dictionary<Entity, Id>;
}

```

And the simpler function we can create to return this shape is:

```ts
function stateFactory<Entity>(): EntityState<Entity> {
  return { ids: [], entities: {} };
}
```

## The actions

Zustand allows the actions to be defined in different ways, using a curried store definition, using `combine`, or placing the actions outside the store. Additionally, Zustand allows partial updates to the `store`. Using this, we can create functions interacting with the EntityState only.

We can define the actions we need with this interface:

```ts

export interface EntityActions<Entity, Id extends EntityId> {
  /**
   * Add one entity to the collection, if that entity doesn't exist already.
   */
  addOne(entity: Entity): void;
  /**
   * Add a list of entities to the collection, if they don't already exist.
   */
  addMany(entities: Entity[]): void;
  /**
   * Add one entity to the collection, replacing any existing entity with the same id.
   */
  setOne(entity: Entity): void;
  /**
   * Add a list of entities to the collection, replacing any existing entity with the same id.
   */
  setMany(entities: Entity[]): void;
  /**
   * Replace all entries in the collection with the list of entities.
   */
  setAll(entities: Entity[]): void;
  /**
   * Applies a partial update in an entity searched by the `id` property, if the entity searched exists.
   */
  updateOne(update: Update<Entity, Id>): void;
  /**
   * Applies a partial update in each entity searched by the `id` property, if the entity searched exists.
   */
  updateMany(updates: Update<Entity, Id>[]): void;
  /**
   * Replaces an entity if the entity exists, or inserts it.
   */
  upsertOne(entity: Entity): void;
  /**
   * Replaces each entity if each entity exists, or inserts it.
   */
  upsertMany(entities: Entity[]): void;
  /**
   * Removes a list of entities from the collection.
   */
  removeMany(entities: Entity[]): void;
  /**
   * Removes an entity from the collection.
   */
  removeOne(entity: Entity): void;
  /**
   * Removes all entities from the collection.
   */
  removeAll(): void;
}
```

To define the factory, we will need three additional functions:

- We need to be able to set the state.
- We need to optionally be able to select a custom id provided by the user.
- We need to optionally be able to sort the list of the entities' ids.

We can define these needs through an options interface.

```ts
interface ActionsFactoryOptions<Entity extends object, Id extends EntityId> {
  idSelector?: IdSelector<Entity, Id>;
  sort?: Comparer<Entity> | false;
  setState: SetState<{}, Entity, Id>;
}
```

Because the `idSelector` is an optional method, and `sort` will be used to compare and not for the actual sorting, we need additional methods to provide a default selector, and to sort the entities.

```ts

function defaultIdSelector<Id extends EntityId>(entity: any): Id {
  if (process.env.NODE_ENV !== 'production' && entity.id === undefined) {
    console.warn(
      'The entity passed to the `selectId` implementation returned undefined.',
      'You should probably provide your own `selectId` implementation.',
      'The entity that was passed:',
      entity
    );
  }

  return entity.id;
}

function getSortedIds<Entity, Id extends EntityId>(
  entities: EntityState<Entity, Id>['entities'],
  idSelector: IdSelector<Entity, Id>,
  sorter: Comparer<Entity>
): Id[] {
  const entitiesList: Entity[] = Object.values(entities);

  entitiesList.sort(sorter);

  const sortedIds = entitiesList.map(idSelector);

  return sortedIds;
}
```  

The store can be updated with a set of independent internal functions, that can be mixed and combined to allow the interaction needed from the action methods.

For example, we can use create a setOne function.

```ts
  const setOne = (state: State, entity: Entity): State => {
    const id = idSelector(entity);

    const entities = {
      ...state.entities,
      [id]: entity,
    };

    const ids = sort
      ? getSortedIds(entities, idSelector, sort)
      : [...state.ids, id];

    return {
      entities,
      ids,
    };
  };
```

And use it to implement multiple actions.

```ts
    setOne(entity) {
      setState((state) => setOne(state, entity));
    },
    setMany(entities) {
      setState((state) => entities.reduce(setOne, state));
    },
    setAll(entities) {
      setState(() => entities.reduce(setOne, EMPTY_ENTITY_STATE));
    },
```

We can define the actionsFactory like this:

```ts

export function actionsFactory<Entity extends object, Id extends EntityId>({
  setState,
  idSelector = defaultIdSelector,
  sort = false,
}: ActionsFactoryOptions<Entity, Id>): EntityActions<Entity, Id> {
  type State = EntityState<Entity, Id>;

  ... // Independent internal functions

  return {
    ... // The actions implementing the interface
  };
}
```

## The selectors

Although Zustand doesn't provide or enforce the usage of selectors, the EntityAdapter generates selectors to read the entities efficiently.

`ngrx` and `Redux Toolkit` have similar selectors, but the Redux Toolkit has an additional selector to query by id an entity. Because of this, my proposal for the selectors is to have this shape:

```ts
export interface EntitySelectors<
  Entity,
  Id extends EntityId,
  State extends EntityState<Entity, Id> = EntityState<Entity, Id>
> {
  selectIds(state: State): EntityId[];
  selectEntities(state: State): Dictionary<Entity, Id>;
  selectAll(state: State): Entity[];
  selectTotal(state: State): number;
  selectById(id: EntityId): (state: State) => Entity | undefined;
  selectById(id: EntityId, state: State): Entity | undefined;
}
```

Notice the `selectById` curried overload, with that we can use the selector if we know the id ahead like:

```ts
const user = useUserStore(selectById(userId));
```

Or use an ID inside the store, like:

```ts
const user = useUserStore(state => selectById(state.selectedUserId, state))
```

To implement that interface we can create a factory like this:

```ts

export function selectorsFactory<
  Entity,
  Id extends EntityId
>(): EntitySelectors<Entity, Id> {
  type State = EntityState<Entity, Id>;

  const selectIds = (state: State) => state.ids;
  const selectEntities = (state: State) => state.entities;
  const selectTotal = (state: State) => state.ids.length;

  const selectAll = ({ entities, ids }: State) =>
    ids.reduce((all, id): Entity[] => {
      const entity = entities[id];
      if (entity) {
        all.push(entity);
      }
      return all;
    }, [] as Entity[]);

  function selectById(id: Id): (state: State) => Entity | undefined;
  function selectById(id: Id, state: State): Entity | undefined;
  function selectById(
    id: Id,
    state?: State
  ): Entity | undefined | ((state: State) => Entity | undefined) {
    if (!state) {
      return (state: State) => selectById(id, state);
    }

    return state.entities[id];
  }

  return {
    selectIds,
    selectEntities,
    selectAll,
    selectTotal,
    selectById,
  };
}
```

## The adapter creator

With all that, I want to propose the following entityAdapter creator:

```ts

interface EntityAdapterOptions<Entity extends object, Id extends EntityId> {
  idSelector?: IdSelector<Entity, Id>;
  sort?: Comparer<Entity>;
}

export function createEntityAdapter<
  Entity extends object,
  Id extends EntityId
>({ idSelector, sort }: EntityAdapterOptions<Entity, Id>) {
  return {
    getState() {
      return stateFactory<Entity, Id>();
    },
    getActions(setState: SetState<Entity, {}, Id>) {
      return actionsFactory({ setState, idSelector, sort });
    },
    getSelectors() {
      return selectorsFactory<Entity, Id>();
    },
  };
}
```

## The store creator

Additionally, I have developed an entity store creator. The implementation I propose can be called with multiple overloads, but it has the same base implementation:

```ts

function createEntityStoreBase<
  Entity extends object,
  ExtraState extends object,
  ExtraActions extends object,
  Id extends EntityId
>(
  { idSelector, sort }: EntityAdapterOptions<Entity, Id> = {},
  stateCreator?: ExtraStateCreator<ExtraState>,
  actionsCreator?: StateCreator<ExtraState, [], [], ExtraActions>
): UseBoundEntityStore<Entity, ExtraState, ExtraActions, Id> {
  type State = StoreState<Entity, ExtraState, Id>;
  type Actions = StoreActions<Entity, ExtraState, ExtraActions, Id>;

  const entityAdapter = createEntityAdapter<Entity, Id>({
    idSelector,
    sort,
  });
  const useStore = create(
    () =>
      ({
        ...entityAdapter.getState(),
        ...stateCreator?.(),
      } as State)
  );

  return Object.assign(useStore, {
    selectors: entityAdapter.getSelectors(),
    actions: {
      ...entityAdapter.getActions(useStore.setState),
      ...actionsCreator?.(useStore.setState, useStore.getState, useStore),
    } as Actions,
  });
}
```

I wrote a couple of posts on dev.to about this:
[Zustand EntityAdapter - An EntityAdapter example for Zustand](https://dev.to/michaeljota/zustand-entityadapter-an-entityadapter-example-for-zustand-cd2)
[Zustand EntityAdapter - A bookshelf example](https://dev.to/michaeljota/zustand-entityadapter-a-bookshelf-example-28bk)

And you can test the implementation in this StackBlitz project.
[Zustand EntityAdapter Proposal](https://stackblitz.com/edit/zustand-entityadapter-proposal-8lzpfj?file=src%2Fzustand-entityadapter%2Fcreators.ts)

**Comment by Wxh16144** [comment]
It seems to have not been taken seriously? I hope to implement this proposal. ❤️

**Reply by dai-shi** [reply]
Thanks for the heads up.
Let's see others think.

This is the field for third-party libraries. I wonder if there's already something similar. Otherwise, feel free to create one.

**Comment by michaeljota** [comment]
I've created a library for those interested in this proposal: https://github.com/michaeljota/zustand-entity-adapter. I hope it is ok to use Zustand in the name, I'm not very creative. This is the npm link: https://www.npmjs.com/package/zustand-entity-adapter.

Although it says it's version 0.1.0, it is stable and I've added tests to cover the expected behavior I consider. So, this should be ready to be used in production, but I'm eager for your feedback to make further improvements.

Thank you all!
